include/oxen/quic/ip.hpp

```diff
@@ -46,13 +47,15 @@ namespace oxen::quic
         ipv4 base;
         uint8_t mask;
 
+        ipv4 max_ip();
+
         const std::string to_string() const { return base.to_base(mask).to_string(); }
 
-        constexpr bool operator==(const ipv4_net& a) const { return base == a.base && mask == a.mask; }
+        constexpr auto operator<=>(const ipv4_net& a) const { return base.to_base(mask) <=> a.base.to_base(a.mask); }
```

Comment:
There's a bit of weirdness here (not added here, though enlarged on by switching from `<` to `<=>`): by only comparing the base address we get weird things like:

auto a = ipv4{10,0,0,0} / 8;
auto b = ipv4{10,1,0,0} / 16;
auto c = ipv4{10,0,0,0} / 16;

a < b; // true
b < a; // false
a < c; // false;
c < a; // false

and yet most of the addresses in the first range are greater than all of the addresses in the second range, so it feels a little weird to say that "a comes before b".

The `a` <-> `c` comparisons are a bigger problem (not just added here, but in the existing code) for std::map or std::set, because those define an element to exist based on whether `!(a < b) && !(b < a)`.  And so if you try to put `a` and `c` in the same map or set, it will fail because, according to this logic, they are equivalent elements.  `std::sort` would have similar obstacles.

I think you *could* implement <=> with [partial ordering](https://en.cppreference.com/w/cpp/utility/compare/partial_ordering): a comes before b if its entire range comes before b's entire range, and likewise for b.  It's not entirely clear to me what that does for `std::map`/`std::set`/`std::sort`, though, but seems like it would be able to encapsulate the relationship here between ranges.  (I'm not sure we actually have a use case for such a thing, but it's an interesting use case for C++20's partial ordering).

Reply:
Yeah it's tricky. Some thoughts:

I see some potential weirdness with ordering based on `a comes before b if its entire range comes before b's entire range, and likewise for b`. I think this is a result of trying to assert equality comparisons between sets that may contain one another. As a result, we're using equality comparison 

Using the example ranges from your comment, the range of possible values of `a`  (10.0.0.0 - 10.255.255.255) encompasses that of `b` (10.1.0.0 - 10.1.255.255), so it would be more accurate to say that `b` is a perfect subset of `a`. 

In that consideration, I see your very valid point about the `a` vs `c` comparison and I think we can address it as follows:

```c++
// explicit constructor ensuring ipv4_net always calls ::to_base(...) on construction outside of the `operator/`
explicit constexpr ipv4_net(ipv4 b, uint8_t m) : base{b.to_base(m)}, mask{m} {}

// modified <=> operator
constexpr auto operator<=>(const ipv4_net& a) const 
{
    return std::tie(base, mask) <=> std::tie(a.base, a.mask);
}
```

Now, `a` and `c` produce the same base IP, but their masks are different; therefore `a` would still come prior to `c` by satisfying the statement `a < c == true`.

Thoughts?

Reply:
Why do we need inequality operators at all?